packages/powersync-op-sqlite/src/db/OPSqliteAdapter.ts

```
@@ -0,0 +1,173 @@
+  }
+
+  writeTransaction<T>(
+    fn: (tx: Transaction) => Promise<T>,
+    options?: DBLockOptions
+  ): Promise<T> {
+    return this.writeLock((ctx) => this.internalTransaction(ctx, fn));
+  }
+
+  getAll<T>(sql: string, parameters?: any[]): Promise<T[]> {
+    return this.readLock((ctx) => ctx.getAll(sql, parameters));
+  }
+
+  getOptional<T>(sql: string, parameters?: any[]): Promise<T | null> {
+    return this.readLock((ctx) => ctx.getOptional(sql, parameters));
+  }
+
+  get<T>(sql: string, parameters?: any[]): Promise<T> {
+    return this.readLock((ctx) => ctx.get(sql, parameters));
+  }
+
+  execute(query: string, params?: any[]) {
+    return this.writeLock((ctx) => ctx.execute(query, params));
+  }
+
+  async executeBatch(
+    query: string,
+    params: any[][] = []
+  ): Promise<QueryResult> {
+    return this.writeLock((ctx) => ctx.executeBatch(query, params));
+  }
+
+  protected async internalTransaction<T>(
+    connection: OPSQLiteConnection,
+    fn: (tx: Transaction) => Promise<T>
+  ): Promise<T> {
+    let finalized = false;
+    const commit = async (): Promise<QueryResult> => {
+      if (finalized) {
+        return { rowsAffected: 0 };
+      }
+      finalized = true;
+      return connection.execute('COMMIT');
+    };
+    const rollback = async (): Promise<QueryResult> => {
+      if (finalized) {
+        return { rowsAffected: 0 };
+      }
+      finalized = true;
+      return connection.execute('ROLLBACK');
+    };
+    try {
+      await connection.execute('BEGIN');
+      const result = await fn({
+        execute: (query, params) => connection.execute(query, params),
+        get: (query, params) => connection.get(query, params),
+        getAll: (query, params) => connection.getAll(query, params),
+        getOptional: (query, params) => connection.getOptional(query, params),
+        commit,
+        rollback,
+      });
+      await commit();
+      return result;
+    } catch (ex) {
+      await rollback();
```

This should rethrow the exception after rollback. Re throwing should remove the requirement for the return statement below.